doc/format-run.md

```diff
@@ -0,0 +1,216 @@
+It follows immediately after the header. The endianness from the header
+indicates the endianness of the 32bit words.
+
+## Index file
+
+The index maps keys to the number of the disk page(s) in the key/operation file
+that _could_ contain this key.
+
+That is, the index says where the key could possiblly be, and thus what pages
+to load from disk to search for the key. The index does not say whether a key
+exists or not (the bloom filter does this, albeit probabilistically, with false
+positives).
+
+For the case of large values that require multiple pages, the index must report
+the span of disk pages that the key/operation may be on. This allows the whole
+range of pages to be read from disk in one go. (An alternative design would be
+for the first page to indicate that it is a multi-page value, and then require
+a second round of disk I/O to read in the remaining pages.)
+
+The design is intended to be somewhat extensible by allowing for different
+choices of index representation, to suite different kinds of key and key
+properties. In particular initially we will support a compact index type.
+For more general purpose applications one can forsee the need for a more
+ordinary index type.
+
+The type of index in use is metadata that must be known for the whole LSM tree.
+It does not vary per run.
+
+### Compact index
+
+The compact index type is designed to work with keys that are large
+cryptographic hashes, e.g. 32 bytes. In particular it requires:
+
+* keys must be uniformly distributed
+* keys must be at least 6 bytes (48bits), but can otherwise be variable length
+
+For this important special case, we can do significantly better than storing a
+whole key per page: we can typically store just 4 bytes (32bits) per page. This
+is a factor of 8 saving for 32 byte keys.
+
+The representation consists of
+1. the number of range finder bits (0..16)
+2. a range finder array, of 2^n entries of 32bit each (n = range finder bits)
+3. a primary array of 32bit words, one entry per page in the index
+4. a clash indicator bit vector, one bit per page in the index
+5. a clash map, mapping each page with a clash indicator to the full minimum
+   key for the page
+
+The file format consists of each part, sequentially within the file. This
+format can in-part be written out incrementally as the index is constructed.
+The primary array is the largest component, and this is the part that can be
+written out to disk incrementally. Its size is bounded but not known precisely
+prior to the completion of the index. The range finder array cannot be written
+until index completion, but its size is known in advance and so space in the
+file can be reserved. The clash indicator bit vector has the same number of
+elements as the primary array, but is of course 32 times smaller, so it is
+reasonable to keep this in memory while the index is constructed and flush it
+upon completion. Similarly, the clash map is small and it can be written upon
+index completion.
```

Comment:
#38 currently outputs both the primary and clashes arrays incrementally, so we should change that in #38 (but it should be a minor change)

Reply:
yes, that should be easy to adjust.